tracer/publisher.go

```diff
@@ -0,0 +1,112 @@
+package tracer
+
+import (
+	"context"
+	"fmt"
+	"log/slog"
+	"math/rand"
+	"net/http"
+	"os"
+	"sync/atomic"
+	"time"
+
+	"nhooyr.io/websocket"
+	"subtrace.dev/pubsub"
+	"subtrace.dev/rpc"
+)
+
+var DefaultPublisher = newPublisher()
+
+type Publisher struct {
+	log atomic.Bool
+	ws  *websocket.Conn
+
+	ch chan *[]byte
+}
+
+func newPublisher() *Publisher {
+	p := &Publisher{ch: make(chan *[]byte)}
+	go func() {
+		if err := p.loop(); err != nil {
+			slog.Error("publish loop failed", "err", err)
+		}
+	}()
+	return p
+}
+
+func (m *Publisher) getWebsocket(ctx context.Context) (*websocket.Conn, error) {
+	if m.ws != nil {
+		return m.ws, nil
+	}
+
+	var attempt int
+	var lastErr error
+	for attempt = 0; attempt < 10; attempt++ {
```

Comment:
try forever because we have no idea how long the connection issue will persist as a fraction of how long servers typically run for